### DISCUSSION TITLE
Needing help with debricking blue Electribe

### DISCUSSION THREAD

**Discussion by nmscoop** [bangcorrupt/hacktribe#305] [Q&A] [answered]
Hey there,

Some time ago I decided to install hacktribe on my E2... Long story short, I didn't pay attention to using the right patch and accidentally used the patch for the sampler. I'm a total newbie when it comes to RPi, OpenOCD or debugging in general but i really want to get my E2 working again. I consider myself clever enough to be able to do achieve this by following the Debrick Guide. (I know that the fact that I wasn't aware of the different patches, isn't really underlining this point... anyway...)
I'm at this point now where I'm running into some issues due to OpenOCD being updated or rather the guide being a bit older.

The first issue to come to my attention was that on the blue Electribe there is **no unpopulated pin header (J4)**. Therefore I couldn't modify the power switch.
(_Do I even need to?_ The screen lights up and stays like that till I pull the plug.
On the other hand: _Does it make a difference, in this context, to shut down the device by pulling the plug (and thus cutting the power supply) or to shut it down but not cutting the power supply_?)

Secondly, the **openocd.cfg config file should be updated**.

    DEPRECATED! use 'bcm2835gpio speed-coefss' not 'bcm2835gpio_speed_coeffs'
    DEPRECATED! use ' adapter gpio tck; adapter gpio tms; adapter gpio tdi; adapter gpio tdo' not 'bcm2835gpio_jtag_nums'
    DEPRECATED! use ' adapter gpio trst' not 'bcm2835gpio_trst_num'

This obviously isn't a big issue but certainly something to be aware of.

I prepared my SD CARD and got the Firmware Binary as I was supposed to.
Now I'm stuck at getting the SBL. My config file is in another directory. When I try to run OpenOCD via `openocd -f ./bootloader/openocd.cfg` it shows the following:

    Open On-Chip Debugger 0.12.0+dev-00440-ge8e09b1b5 (2023-12-23-15:53)
    Licensed under GNU GPL v2
    For bug reports, read http://openocd.org/doc/doxygen/bugs.html
    force hard breakpoints
    Info : BCM2835 GPIO JTAG/SWD bitbang driver
    Info : clock speed 1496 kHz
    Info : JTAG tap: omapl138.jrc tap/device found: 0x1b7d102f (mfg: 0x017 (Texas Instruments), part: 0xb7d1, ver: 0x1)
    Info : JTAG tap: omapl138.etb enabled
    Info : JTAG tap: omapl138.arm enabled
    Info : Embedded ICE version 6
    Info : omapl138.arm: hardware has 2 breakpoint/watchpoint units
    Info : ETM v1.3
    Info : [omapl138.arm] Examination succeed
    Info : starting gdb server for omapl138.arm on 3333
    Info : Listening on port 3333 for gdb connections
    Warn : target was in unknown state when halt was requested target halted in ARM state due to debug-request, current mode: Undefined instruction
    cpsr: 0x200000db pc: 0xc0000138
    MMU: enabled, D-Cache: enabled, I-Cache: enabled

I don't know what any of that means. But what I know is that **OpenOCD doesn't open up the ports**, it's supposed to. Therefore **I cannot connect to the telnet server** (which would be the next step after running OpenOCD).

Any bit of help would be appreciated. I'm hoping to get my E2 fixed (soon but we'll see how long it'll take).

Thanks, scoop.

**Comment by bangcorrupt** [comment]
> accidentally used the patch for the sampler

The patch is the same, only the file header changes.  If the header is wrong, the device should refuse to install.  How did you make it? 

Can you email me the file you installed?  It won't help you debrick, but it can help identify errors in the tools and documentation. 


> no unpopulated pin header (J4)

That's interesting, it's not on the red one either.  You will be able to dump the bootloader without modifying the power switch, but completing the whole process manually holding the power button while typing will be error prone.  There should be other points for connecting the switch, I'm not sure if J4 is connected directly across the power button, I will check.  

> openocd.cfg config file should be updated

You're right that this shouldn't cause an issue, but if you open a PR I'll merge it once I've tested.

> I don't know what any of that means

The openocd log shows it connects correctly to the CPU, which has halted at address 0xc0000138.  It is listening for GDB connections on port 3333, but you're right that it doesn't appear to be listening on port 4444 for telnet connections.  


I don't have everything set up right now, but I will try to debug this over the next few days.  Some guesses:

Have you edited the config file at all?

Are you connecting to the telnet server from the same host it is running on?  

Are you using a firewall?

Do you have anything else listening on 4444?  

While openocd is still running, run in another terminal `ss -tunap | grep 4444`, this will show any processes bound to port 4444.

I would have thought it would print an error message if it tried to bind the port and failed.  Openocd should start a telnet server on 4444 unless told otherwise:

https://openocd.org/doc/html/Server-Configuration.html

See section 7.3 TCP/IP Ports


EDIT:

You can run openocd with `-d3` to print a lot more information.

Also try running both openocd and telnet with root privileges to check if it is a permission issue.


EDIT 2:

If you really can't get telnet working, you should be able to use the GDB server instead, either using it to issue commands to openocd, or using gdb to set breakpoints and write memory.

From the GDB prompt run `targ ext localhost:3333` to connect to the openocd GDB server.

Then prefix openocd commands with `monitor` to send them through to openocd, for example:

    monitor dump_image SBL.bin 0x80000000 0x5e40

**Reply by nmscoop** [reply]
Hey @bangcorrupt!
It's been over 9 months since I initially tried to debrick my Electribe. But I had things to do and forgot about that... Therefore I also want to tell you that I'm sorry to not have responded. Either way, with your new input I finally thought to give it a new try.

_FYI, I still don't have a clue about OpenOCD, GDB or debugging in general. Maybe I'm a bit too ambitious with this. But I don't want to give up that easily.
Also I had to factory reset my RPi because my SD Card was corrupted. (Thus, my RPi/the Raspberry Pi OS has it's factory settings.)_

## Patch File & Power Switch
First up, I can't provide you the same exact patch file anymore because I deleted it basically right after I bricked it or rather right before I tried to debrick it. But if the device would refuse to install the wrong patch file then I'm guessing that I was just impatient and shut off the device too early or something.
And regarding the additional power switch: I still don't have any idea where I may be able to connect a new power switch. Did you maybe find anything?


## OpenOCD Config File
Secondly, the openocd config file meanwhile contains some new deprecations. And I'm telling you that because you asked if I had changed the config file at all. Yes, I updated it so it doesn't contain the depracted command lines anymore and also I changed the source path because I imported the config files into another directory. But that shouldn't cause an issue, right? (Just to make sure, here's my config file:)

```
adapter driver bcm2835gpio

bcm2835gpio peripheral_base 0x20000000

bcm2835gpio speed_coeffs 113714 28

adapter gpio tck 11
adapter gpio tms 25
adapter gpio tdi 10
adapter gpio tdo 9

adapter gpio trst

transport select jtag
set CHIPNAME omapl138

source ./bootloader/am1802.cfg

gdb memory_map enable
gdb flash_program enable

reset_config trst_and_srst
reset_config trst_push_pull

init

halt 1000

adapter speed 0
```

## Your New Input

**But let's get to the real problem:** Even with your new input I'm still stuck where I left off last time. OpenOCD's warning message is exactly the same. But at least this time I can attach the output of `openocd -f ./bootloader/openocd.cfg -d3`: [d3.txt](https://github.com/user-attachments/files/16404986/d3.txt) (I don't actually know if this is of any use, but maybe you're able to see something that I don't.)

I also figured, that I'm missing some steps in order to get a telnet connection running. I made sure everything is installed and such but when I try to run `telnet localhost 4444` it refuses the connection:

```
admin@raspberrypi:~ $ telnet localhost 4444
Trying ::1...
Connection failed: Connection refused
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection refused
```

Am I missing something here? Yes, I'm connecting from the same host it is running on. Though I'm a bit confused on wether "connecting" in this context means running `telnet localhost 4444` or if I'm overseeing something. Do I have to run any other command in order to connect to telnet before running OpenOCD?

By the way, no I'm not using a firewall. And no, I don't have anything else listening on 4444.

> If you really can't get telnet working, you should be able to use the GDB server instead, either using it to issue commands to openocd, or using gdb to set breakpoints and write memory.
>
> From the GDB prompt run `targ ext localhost:3333` to connect to the openocd GDB server.

I tried. But when running `targ ext localhost:3333` nothing happens and after a short time it outputs `localhost:3333: Connection timed out`. What am I doing wrong here?

## What am I even doing?

From the bottom of my heart, thank you for your time and patience with me. I know, it seems very stupid of me to take on this challenging project with absolutely no knowledge whatsoever. Should've paid more attention in linux class in my informatics bachelor, ig... Also I really try to find answers for myself but this is such a specific problem, I feel like, that it is hard to gather any useable information - and I don't even mean understanding this information yet.
After 9 months your help is once again very appreciated!

Thanks, scoop.

**Reply by bangcorrupt** [reply]
> Power switch

You should be able to solder the switch across the pins of the power button.  I haven't actually tested this, but it would be the same as holding the button down, so I can't see any issues.

> Config File...

Your log files show that OpenOCD connects to the processor correctly, so your config file must be correct.  OpenOCD should start the telnet server automatically, but the log file shows it does not.  It also shows that the GDB server is started, but for some reason you can't connect to it.

My first thought is that this is a permissions issue with your user account.  You can check this by prefixing all the commands with `sudo` to run them with root permissions.  If it works with `sudo` you can look into how to give your user account the correct privileges.

If you are already logged in as root, or it doesn't work with `sudo` then I don't know why you can't start the telnet server, or connect to the GDB server.  

> What am I even doing?

Don't worry about it, this is hard.  It took me ages to make it work.  The good news is the connection between OpenOCD and the processor is working, so you only need to fix Linux issues, not the hardware.

Which Linux distro are you using on the rpi?  Is it the default installation, or does it have a lot of configuration changes?

EDIT:

I just thought, check that you have `telnetd` package installed.  I think the client and server are separate packages and may not both be installed by default.

**Reply by nmscoop** [reply]
I cannot stress this enough: **thanks for your help (and your quick reply) and thanks for the kindness and patience with me!**

It's good to hear that it should not be an hardware related problem. I'll try to look more into the whole OpenOCD/GDB thing myself. I haven't actually tried to run everynthing with root permission... I hope that it'll fix it.

I'm pretty sure that I checked to have every necessary package to be installed, even `telnetd`, but I will make sure to check again.

As I've said, I'm not expert when it comes to RPi/Linux, but as far as I can say I didn't change anything configuration-wise. I'm using Raspberry Pi OS (Raspbian) as already mentioned in my previous reply and I basically installed the newest version with standard configuration just with a different username. That's all.

Hopefully my next reply will be the last, in which I'm telling you that I was successfull. We'll see. For now, I'm going to mark this as the answer.

**Comment by andresblack** [comment]
Hi, i´m on the same issue. Was reading and try all the sugested solutions but i´m stuck.
RPI3, Fresh install 64bit OS.
Modified both .cfg files but i can´t telnet also.
When run openocd i get this:
Open On-Chip Debugger 0.12.0+dev-00825-g250ab1008 (2025-01-06-07:04)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
force hard breakpoints
Info : BCM2835 GPIO JTAG/SWD bitbang driver
Info : clock speed 1496 kHz
Info : JTAG tap: omapl138.jrc tap/device found: 0x1b7d102f (mfg: 0x017 (Texas Instruments), part: 0xb7d1, ver: 0x1)
Info : JTAG tap: omapl138.etb enabled
Info : JTAG tap: omapl138.arm enabled
Info : Embedded ICE version 6
Info : omapl138.arm: hardware has 2 breakpoint/watchpoint units
Info : ETM v1.3
Info : [omapl138.arm] Examination succeed
Info : [omapl138.arm] starting gdb server on 3333
Info : Listening on port 3333 for gdb connections
Warn : target was in unknown state when halt was requested
target halted in ARM state due to debug-request, current mode: Undefined instruction
cpsr: 0x600000db pc: 0xc00002ac
MMU: enabled, D-Cache: enabled, I-Cache: enabled

**Reply by bangcorrupt** [reply]
> Info : JTAG tap: omapl138.jrc tap/device found: 0x1b7d102f (mfg: 0x017 (Texas Instruments), part: 0xb7d1, ver: 0x1)

This shows it has found the correct processor.

> target halted in ARM state due to debug-request, current mode: Undefined instruction
> cpsr: 0x600000db pc: 0xc00002ac

This shows it has crashed trying to execute from an address that does not contain a valid instruction.

The debugger is connected correctly and OpenOCD is working, but the telnet server is not starting.  

>  check that you have telnetd package installed. I think the client and server are separate packages and may not both be installed by default.

The client is `telnet` and the server is `telnetd`, what happens if you run `telnetd --version` ?

**Comment by andresblack** [comment]
| This shows it has crashed trying to execute from an address that does not contain a valid instruction.

Is this normal or maybe a dupont cable or E2 hardware issue?

| The client is telnet and the server is telnetd, what happens if you run telnetd --version ?

root@rpi:/home/administrator/openocd# telnetd --version
telnetd (GNU inetutils) 2.4
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by many authors.

**Reply by bangcorrupt** [reply]
This is normal, it's why it is bricked.  The log shows the hardware is connected correctly and OpenOCD is working.

I have no idea why it doesn't start a server, the documentation says it should start automatically:

https://openocd.org/doc-release/html/Running.html#What-OpenOCD-does-as-it-starts

It may be worth asking on the OpenOCD mailing list:

https://openocd.org/pages/discussion.html

I think this is probably a Linux configuration issue, but I'm out of ideas why.

Running `openocd` and `telnet` with debugging enabled could provide more information:

```
openocd -d3

telnet -d localhost 4444
```

**Comment by andresblack** [comment]
Any way to get older openocd version like te one you used and its known to work?

**Comment by andresblack** [comment]
I got it working!!!
Using the xpack-openocd as in this tutorial:
https://docs.google.com/document/d/1DzCMbvGbJhT4pLrXS6tPKS1AzACUtOUoZ7YgnNTKP4w/edit?tab=t.0

**Reply by bangcorrupt** [reply]
Excellent, I'll add a link to this in the wiki.

**Comment by nmscoop** [comment, marked as answer]
After nearly 3 years of trying to debrick my Electribe, i finally got it.

And the insane part is that my cables from GPIO to JTAG were the problem all along. They prevented OpenOCD from opening the telnet port because the connection was instable all the time. OpenOCD couldn't halt in ARM-state, which is required in order to open telnet and tcl ports afterwards.

I adjusted the cables (as well as the clocking speed (adapter speed from 0 to 50) in the config file) and got it working after that.
Holy shit, I can't believe it.

**Reply by bangcorrupt** [reply]
> OpenOCD couldn't halt in ARM-state, which is required in order to open telnet and tcl ports afterwards.

Looking back, that's probably what this warning was trying to tell us:

> Warn : target was in unknown state when halt was requested target halted in ARM state due to debug-request, current mode: Undefined instruction

I spent nearly a year figuring this out initially, most of that was probably due to noisy connections.

**Reply by vanana-old-acc** [reply]
> After nearly 3 years of trying to debrick my Electribe, i finally got it.
> 
> And the insane part is that my cables from GPIO to JTAG were the problem all along. They prevented OpenOCD from opening the telnet port because the connection was instable all the time. OpenOCD couldn't halt in ARM-state, which is required in order to open telnet and tcl ports afterwards.
> 
> I adjusted the cables (as well as the clocking speed (adapter speed from 0 to 50) in the config file) and got it working after that. Holy shit, I can't believe it.

I get a similar warning:
`target halted in ARM state due to debug-request, current mode: Supervisor`
Note it doesn't say Undefined instruction, but Supervisor mode.
That would indicate noisy connections?

**Reply by bangcorrupt** [reply]
> `target halted in ARM state due to debug-request, current mode: Supervisor`

This looks like what you're supposed to get, you should be able to connect GDB now.

If you post the full log back in the other thread I can help you work it out.  #393
